lib/streamlit/widgets.py

```diff
@@ -73,15 +75,60 @@ def get_widget_value(self, widget_id: str) -> Optional[Any]:
 
     def set_state(self, widget_states: WidgetStates) -> None:
         """Copy the state from a WidgetStates protobuf into our state dict."""
-        self._state = {}
+        self._widget_states = {}
         for wstate in widget_states.widgets:
-            self._state[wstate.id] = wstate
+            self._widget_states[wstate.id] = wstate
+
+    def add_callback(
+        self,
+        widget_id: str,
+        deserializer: Callable[..., Any],
+        callback: Callable[..., None],
+    ) -> None:
+        """Add a callback that will be called immediately before the app's
+        next rerun if the given widget's value has changed.
+        """
+        self._widget_callbacks[widget_id] = callback
+        self._widget_deserializers[widget_id] = deserializer
+
+    def get_callback(self, widget_id: str) -> Optional[Callable[..., None]]:
+        return self._widget_callbacks.get(widget_id, None)
+
+    def call_callbacks(self, new_widget_states: WidgetStates) -> None:
+        """Call all callbacks based on changes in the widget state. We are
+        assuming that widgets send the same state value each time if unchanged.
+        """
+        for new_state in new_widget_states.widgets:
+            callback = self._widget_callbacks.get(new_state.id, None)
+            deserializer = self._widget_deserializers.get(new_state.id, None)
+            if callback is None:
+                # The widget does not have an on_change callback.
+                continue
+
+            old_value = self.get_widget_value(new_state.id)
+            new_value = _get_widget_value(new_state)
+
+            if deserializer is not None:
+                old_value = deserializer(old_value)
+                new_value = deserializer(new_value)
+
+            if new_value == old_value:
+                # The widget's value has not changed.
+                continue
```

Comment:
I _think_ this is fine, but I think it also means that deserializer values need to be `__eq__`'alable with each other (so if your deserializer returns some crazy-ass custom type, you need to ensure it can be compared with another instance of itself).

Another approach - I'm not necessarily saying I think this is the right way to go - is to do the `if new_value == old_value` equality check _before_ the deserialization step, so that we're just relying on widget protobuf values which we know are equatable. (I think this approach would have fewer code landmines in it but might also break some use cases...?)

Reply:
Yea. One of the biggest problems is that the initialization of the widgets usually return `None`, so oftentimes actually, the value may not change...but what gets sent back is a different value. I had to deserialize to fix that. That being said. I did realize a "bug-ish" in the deserializer for the text inputs that showed a different value, so it's possible it's a moot point. I'll do one more pass.